### ISSUE TITLE
request: [v] doesn't bang a value if it is created after the value is sent

### ISSUE THREAD

**Issue by AydinHumphries** [pure-data/pure-data#1167]
using OS windows 10, lenovo

I created many examples where this happens but in a nutshell, this issue is simple:

if you send a value, e.g. 5 to [send message] then create [value message] and bang it, the float doesn't appear. I think it would be advantageous to have the float appear for a number of reasons. 

anyway the examples/steps to recreate can be found here:

[v-error.zip](https://github.com/pure-data/pure-data/files/5306834/v-error.zip)

**Comment by AydinHumphries** [comment]
example number three was not working when I tried it earlier but after downloading my own patch it seems to work, so there is some inconsistency, it seems. maybe because of the way I've dynamic patched it without iemguts. but this is definitely an issue independent of dynamic patching.

**Comment by porres** [comment]
you can't expect an object that doesn't exist yet to receive a message, as it doesn't exist. This is the same as sending a message to a receive object that doesn't exist.

**Comment by porres** [comment]
Therefore, not a bug or an issue, just a wrong expectation from the user

**Comment by AydinHumphries** [comment]
> Therefore, not a bug or an issue, just a wrong expectation from the user

ok but if I send a message, "message" to [v message] in an abstraction that has yet to be created, where [v message] has to be made because of different messages received, it seems like a unique situation where such a feature could be useful.

the object hasn't been created yet but surely there could be a feature where it would be useful to create an object, somewhere in the Pd source code, after sending an object.

in other words, it's not a bug report but a feature request, essentially. I think it would be useful, if v could bang messages from a message already sent without the v object having been created yet.

**Comment by Spacechild1** [comment]
> I think it would be useful, if v could bang messages from a message already sent without the v object having been created yet.

Seriously, how should that work? That's impossible per definition.

> I think it would be useful, if v could bang messages from a message already sent without the v object having been created yet.

You simply can't send a message to an object that hasn't been created yet. Just make sure that the `[v message]` object exists *somewhere* before sending a message to it.

**Comment by AydinHumphries** [comment]
> > I think it would be useful, if v could bang messages from a message already sent without the v object having been created yet.
> 
> Seriously, how should that work? That's impossible per definition.


Ok so I have limited knowledge about source code. but I have a feeling the code could be used to implement an object that's not visible on the patch but "exists" in the Pd memory after sending a message. I can't prove this is possible but I think it is.


> You simply can't send a message to an object that hasn't been created yet. Just make sure that the `[v message]` object exists _somewhere_ before sending a message to it.

in my patch that wouldn't have been possible before I found a way round this. the [v] object was in an abstraction that had to be dynamically created for an instance of it. the bang to [v] was sent only through the event I changed something on the main patch after the abstraction was created. obviously this requires a convoluted explanation, I just think such a feature could be useful.

**Comment by porres** [comment]
> I can't prove this is possible but I think it is.

again, it's not possible to send things to things that do not exist because they don't exist. This can be proved, in fact, your example proves it already, another proof is:
1) Create a new blank patch;
2) Create an object called [send $0-x];
3) Do not create a [receive $0-x]
4) Send a message with [send]
5) See that there's no  object to receive the message and the message doesn't go anywhere
6) Create  [receive $0-x] (now it exists, but it has no prior knowledge to what happened before its existence and cannot receive messages that were sent before it existed)

> Ok so I have limited knowledge about source code

we do, and that's how it works, it's the same thing.

> I just think such a feature could be useful

It would require a very convoluted and weird change to the code to allow something like this to happen, not to mention it'd be just bad for a few reasons. And I'm pretty sure you can better manage your edge use case to make it work without this feature.

**Comment by fdch** [comment]
More telling is what happens if you open pd and send a ";foo 10" message
with the File>Message: you get a nice, this-is-not-possible red error on
console: "foo: no such object".

**Comment by umlaeute** [comment]
> >    Seriously, how should that work? That's impossible per definition.

> Ok so I have limited knowledge about source code. but I have a feeling the code could be used to implement an object that's not visible on the patch but "exists" in the Pd memory after sending a message.


just to chime in:

theoretically it would of course be possible to store the values, even with no `[value]` object existing yet.

however, that would mean that for each and every send-label we must store the (last) value sent, as somebody might want to access it later.

now i agree with @Spacechild1 & @porres that this is quite a bit of an overkill.

1. the standard mode of operation for Pd is to first create a patch, and then run it. now we all "run" our patches while we build them up, and some of us (your sincerely included) do live-coding on stage. but that doesn't change that the standard mode of operation is otherwise. having `[send]` store values in not-yet-existing `[value]` instances is *only* useful if you are still building up your patch. it's totally useless if the patch is finished before you start `[send]`ing.

1. the semantics of storing a sharing value are embedded in the `[value]` object, not in the `[send]` object. thus you can only share the value if you already have a `[value]` object (in a way this is just rephrasing what my previous speakers already said; they key part, however, is the *semantics* which should explain *why* it is not a good idea)

1. but because it's the death day of james dean, here's a little present. use it as a drop-in replacement for `[send]` and marvel at the powers of abstractions:

~~~
#N canvas 6 87 243 172 12;
#X obj 104 42 inlet;
#X obj 104 67 s \$1;
#X obj 104 92 v \$1;
#X connect 0 0 1 0;
~~~

**Comment by porres** [comment]
yeah, that's a good way to better manage this as I assumed :)

**Comment by Spacechild1** [comment]
Or: why not simply do `[123( -> [v message]`?

**Comment by AydinHumphries** [comment]
> Or: why not simply do `[123( -> [v message]`?

yeah substitute [s] for [v], why did I not think of that?

in any case (though it hardly matters now) I do believe I was right that it is possible to send a message to an object that doesn't exist. 

in my own patch, the abstractions have [v] inside. 

the abstraction:

![image](https://user-images.githubusercontent.com/48067862/94933039-bf52b580-04c1-11eb-951e-a6691bdbb622.png)

the first time creating an abstraction the v does not send: 

![image](https://user-images.githubusercontent.com/48067862/94933071-caa5e100-04c1-11eb-87da-5281d0c7a2d4.png)

but you can do it a second or third time, etc. and it works, even though those objects have just been created and under normal circumstances wouldn't send messages:

![image](https://user-images.githubusercontent.com/48067862/94933172-f1fcae00-04c1-11eb-828f-fe9b2c5346d5.png)

so it seems there is _some_ inconsistency with the coding even if there are a whole host of ways round it.

**Comment by AydinHumphries** [comment]
> it's totally useless if the patch is finished before you start [send]ing.

that's not completely true. if you need to send message to an abstraction that has to be created dynamically is an instance I already mentioned where it could be useful to store a value for an object that has yet to be created. but there are other ways round it as mentioned.

**Comment by porres** [comment]
> I do believe I was right that it is possible to send a message to an object that doesn't exist.

no it isn't

> the first time creating an abstraction the v does not send

because you sent to an object that doesn't exist

> but you can do it a second or third time, etc. and it works

yeah, because you created the object on the first time, so now the thing exists and you can send things to it...

but this doesn't mean it is possible to send a message to an object that doesn't exist. On the contrary,  it only proves the point we've been telling you here that it is not possible to send a message to an object that doesn't exist.

**Comment by AydinHumphries** [comment]
> yeah, because you created the object on the first time, so now the thing exists and you can send things to it...

it's not the same thing however. to demonstrate, I bang [2{ into [s message] ([v message] does not yet exist):

![image](https://user-images.githubusercontent.com/48067862/94952180-790b4f80-04dd-11eb-81a1-a1f9a28a5041.png)

now I create [v message] and bang [v message] and nothing:

![image](https://user-images.githubusercontent.com/48067862/94952265-9c35ff00-04dd-11eb-8dc2-9c8563901883.png)

if I create another [v message] it shouldn't matter if it is in an abstraction or the main patch, in theory it should do the same thing (nothing). so I create a second instance of [v message] on the main patch and obviously, nothing:

![image](https://user-images.githubusercontent.com/48067862/94952366-c5ef2600-04dd-11eb-8ad2-23a4d8ee6538.png)

I create an abstraction and still, nothing:

![image](https://user-images.githubusercontent.com/48067862/94952507-ffc02c80-04dd-11eb-8384-27f07b0cd0c9.png)

**Comment by AydinHumphries** [comment]
> but this doesn't mean it is possible to send a message to an object that doesn't exist. On the contrary, it only proves the point we've been telling you here that it is not possible to send a message to an object that doesn't exist.


I suppose you are saying if I create [v message] and then bang [2{ in [s message] a second time then create [v message] it will send but then why the second time and not the first time? the second instance I created [v message] I still had not created that particular object before I banged [2{. it's still a new object, there just happens to be an old version of that object which had not received [2{ before it was sent either.

**Comment by AydinHumphries** [comment]
also, I can't prove this but in the patch I was working, before I found a way around this [v] issue, the float would not send to a particular [v] object that was non-existent in an abstraction yet to be created but there were three other [v] objects yet to be created in that abstraction that when they were, they banged the values sent just fine.

**Comment by AydinHumphries** [comment]
> I suppose you are saying if I create [v message] and then bang [2{ in [s message] a second time then create [v message] it will send but then why the second time and not the first time? the second instance I created [v message] I still had not created that particular object before I banged [2{. it's still a new object, there just happens to be an old version of that object which had not received [2{ before it was sent either.

it's possible this works after perhaps because I banged [2{ the second time and then it didn't matter that it was a new [v message] because one other already existed. but that particular object didn't exist. so why is it possible to store a value as part of a message provided one object with that message received exists for another that doesn't yet exist but not for an object that does not yet exist. surely it has got to be possible to store a float or symbol with a sent message before an object devised for receiving that object exists, it just has to. source code has a huge memory: it will remember the fact a message was sent and what was contained within that message, regardless if that message has yet been received.

**Comment by porres** [comment]
well, in the 2nd time you sent a value to a [v message] object you had already created an object named [v message], so it receives the value...

when you create new [value] object with the same "message" name, it loads the current value the variable "message" has, if any. In your case, it had one... 

It's still impossible to send values to non existing objects. So just don't do that, find a way to create the object first, it's quite simple and makes total logical sense. And you were already provided with a simple solution for that anyway, one that doesn't need to change the source code of Pd to implement something that doesn't make logical sense (allowing the user to send things to things that don't exist yet).

**Comment by Spacechild1** [comment]
@AydinHumphries to be blunt: I think you don't really understand how `[value]` works...

To expand @porres's explanation: all `[value]` objects of the same name point to the same variable. The first `[value]` object of a given name actually creates the variable. If you delete all `[value]` instances of a given name, the variable is also deleted.

Generally, the issue tracker is not a place to teach Pd... We made it clear early on that this is a user error, so you could do one or more of the following things:
a) read the documentation
b) ask on the Pd mailing list
c) ask on the Pd Facebook group
d) ask on the Pd forum (https://forum.pdpatchrepo.info/)

**Comment by AydinHumphries** [comment]
> And you were already provided with a simple solution for that anyway, one that doesn't need to change the source code of Pd to implement something 

Yeah I know, I said as much.

**Comment by porres** [comment]
> as for reading the help files, I don't think intricate information about Pd source code is available there.

Of course intricate information about Pd's source code is not there, but this is not an intricate information about Pd's source and the relevant information is there and clear, have you read it anyway? It says that all [value] objects with the same name share (thus load) the same value. It doesn't say that others objects won't load the value as you assumed (on the contrary!) or anything about you needing to create all objects before you set values so they can share the value. This is as clear as can be at least in the example where you have a [value] object and you can change its name to access different variable names, as you have an object loading a new variable name (which is like creating a new object with that variable name) and then you can access a value that was set before it loaded that variable name - so it's all about a simple deduction if you read it and didn't get it.

We also tried to make it clear early on, most of this was just us trying to explain you why and I agree this is not the place for it. Even if you've read the documentation and did not get it, you can ask for more help getting it on the mailing list, the forum or the facebook group. This is also not the place to talk about how you did not get the documentation, we can do this on facebook like we've actually done before without problems. You can, though, suggest to change the documentation here if you think it is flawed.
